File: mobile/lib/features/sales/domain/sale.dart
enum SaleStatus { completed, refunded, partial }

class Sale {
  Sale({
    required this.id,
    required this.totalAmount,
    required this.saleType,
    required this.createdAt,
    this.customerId,
    this.customerName,
    this.status = SaleStatus.completed,
    this.items = const [],
    this.payments = const [],
    this.note,
  });

  final String id;
  final double totalAmount;
  final String saleType; // CASH, CREDIT, MIXED
  final DateTime createdAt;
  final String? customerId;
  final String? customerName;
  final SaleStatus status;
  final List<SaleItem> items;
  final List<SalePayment> payments;
  final String? note;

  factory Sale.fromMap(Map<String, dynamic> map) {
    return Sale(
      id: map['id'] as String,
      totalAmount: (map['total_amount'] as num).toDouble(),
      saleType: map['sale_type'] as String? ?? 'CASH',
      createdAt: DateTime.parse(map['created_at'] as String),
      customerId: map['customer_id'] as String?,
      customerName: map['customer_name'] as String?,
      status: _parseStatus(map['status'] as String?),
      note: map['note'] as String?,
    );
  }

  static SaleStatus _parseStatus(String? s) {
    switch (s) {
      case 'refunded':
        return SaleStatus.refunded;
      case 'partial':
        return SaleStatus.partial;
      default:
        return SaleStatus.completed;
    }
  }
}

class SaleItem {
  SaleItem({
    required this.id,
    required this.productId,
    required this.productName,
    required this.qty,
    required this.unitPrice,
  });

  final String id;
  final String productId;
  final String productName;
  final double qty;
  final double unitPrice;

  double get lineTotal => qty * unitPrice;

  factory SaleItem.fromMap(Map<String, dynamic> map) {
    return SaleItem(
      id: map['id'] as String,
      productId: map['product_id'] as String,
      productName: map['product_name'] as String? ?? '',
      qty: (map['qty'] as num).toDouble(),
      unitPrice: (map['unit_price'] as num).toDouble(),
    );
  }
}

class SalePayment {
  SalePayment({
    required this.id,
    required this.method,
    required this.amount,
  });

  final String id;
  final String method; // CASH, QR, BANK, CREDIT
  final double amount;

  factory SalePayment.fromMap(Map<String, dynamic> map) {
    return SalePayment(
      id: map['id'] as String,
      method: map['method'] as String? ?? 'CASH',
      amount: (map['amount'] as num).toDouble(),
    );
  }
}

File: backend/app/schemas/sale.py
from datetime import date, datetime
from decimal import Decimal
from enum import Enum

from pydantic import BaseModel, Field


class SaleType(str, Enum):
    CASH = "CASH"
    CREDIT = "CREDIT"
    MIXED = "MIXED"


class PaymentMethod(str, Enum):
    CASH = "CASH"
    QR = "QR"
    BANK = "BANK"
    CREDIT = "CREDIT"


class SaleItemCreate(BaseModel):
    product_id: str
    qty: Decimal = Field(gt=0)
    unit_price: Decimal = Field(gt=0)


class SaleCreate(BaseModel):
    sale_type: SaleType
    payment_method: PaymentMethod | None = None
    customer_id: str | None = None
    items: list[SaleItemCreate]
    payments: list["SalePaymentCreate"] | None = None


class SaleItemOut(BaseModel):
    id: str
    sale_id: str
    product_id: str
    qty: Decimal
    unit_price: Decimal
    line_total: Decimal

    model_config = {"from_attributes": True}


class SaleOut(BaseModel):
    id: str
    store_id: str
    sale_type: str
    payment_method: str | None = None
    customer_id: str | None
    total_amount: Decimal
    sale_date_ad: date | None = None
    created_at: datetime
    updated_at: datetime
    items: list[SaleItemOut]
    payments: list["SalePaymentOut"] = []

    model_config = {"from_attributes": True}


class SaleListResponse(BaseModel):
    items: list[SaleOut]
    total: int
    page: int
    page_size: int


class SalePaymentCreate(BaseModel):
    method: PaymentMethod
    amount: Decimal = Field(gt=0)


class SalePaymentOut(BaseModel):
    id: str
    sale_id: str
    method: str
    amount: Decimal
    created_at: datetime

    model_config = {"from_attributes": True}


class SaleRefundItemCreate(BaseModel):
    product_id: str | None = None
    sale_item_id: str | None = None
    qty: Decimal = Field(gt=0)


class SaleRefundCreate(BaseModel):
    items: list[SaleRefundItemCreate] | None = None
    reason: str | None = None


class SaleRefundItemOut(BaseModel):
    id: str
    refund_id: str
    sale_id: str
    product_id: str
    qty: Decimal
    unit_price: Decimal
    line_total: Decimal

    model_config = {"from_attributes": True}


class SaleRefundOut(BaseModel):
    id: str
    store_id: str
    sale_id: str
    amount: Decimal
    refund_date_ad: date | None = None
    reason: str | None
    created_at: datetime
    items: list[SaleRefundItemOut]

    model_config = {"from_attributes": True}
